Add reading NNCF config from checkpoint meta in test.py and export.py

LeonidBeynenson opened the pull request.

LeonidBeynenson committed:
Add reading NNCF config from checkpoint meta in test.py and export.py

Ilya-Krylov commented on mmdet/integration/nncf/compression.py:
```diff
@@ -61,6 +64,30 @@ def is_checkpoint_nncf(path):
+            'nncf_config': cfg.get('nncf_config'),
+            'find_unused_parameters': cfg.get('find_unused_parameters')
+            }
+    if nncf_config_part.get('nncf_config', {}).get('log_dir'):
```
looks like `'nncf_config'` is always present in `nncf_config_part`

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

LeonidBeynenson committed:
Fix issue with alt_ssd_export if trying to read NNCF config from checkpoint meta

Ilya-Krylov commented on mmdet/integration/nncf/compression.py:
```diff
@@ -61,6 +64,30 @@ def is_checkpoint_nncf(path):
     except FileNotFoundError:
         return False
 
+def get_nncf_config_from_meta(path, log_dir=None):
```
I guess it is worth removing `log_dir` from here

LeonidBeynenson replied:
I think it should be a good idea in the future, so I'd like to keep it. 
But if your opinion is to remove it definitely, I will remove it

Ilya-Krylov replied:
don't care about future so much, it is not used for now, let's remove it

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

LeonidBeynenson committed:
Apply isort changes

Ilya-Krylov commented on tools/export.py:
```diff
@@ -204,6 +205,14 @@ def main(args):
     fake_data = get_fake_input(cfg, device=device)
 
     # BEGIN nncf part
+    alt_ssd_export=getattr(args, 'alt_ssd_export', False)
```
spaces

LeonidBeynenson replied:
Fixed

LeonidBeynenson committed:
Fix comments from PR

LeonidBeynenson committed:
Decorative changes

druzhkov-paul commented on tools/export.py:
```diff
@@ -204,6 +207,14 @@ def main(args):
     fake_data = get_fake_input(cfg, device=device)
 
     # BEGIN nncf part
+    alt_ssd_export = getattr(args, 'alt_ssd_export', False)
```
```suggestion
    alt_ssd_export = args.alt_ssd_export
```
??

LeonidBeynenson replied:
No, I think we cannot do that. The reason is that we have two subparsers, for `openvino` and for `onnx`, and `onnx` subparser does not contain `alt_ssd_export` parameter.

Please, see below in the function `parse_args`

But this code is called both for `openvino` and `onnx` cases.

druzhkov-paul replied:
Ok, I see. This way 
```suggestion
    alt_ssd_export = args.target == 'openvino' and args.alt_ssd_export
```
looks more explicit and transparent.

LeonidBeynenson replied:
Pavel, note that I just copied the code `getattr(args, 'alt_ssd_export', False)` from several lines below. 
And I think that just check if `alt_ssd_export` is in attributes is more pythonic here...

But if you relly insist on this change, I will apply it

druzhkov-paul replied:
Guess who is the author of the code which is several lines below 😉 I have no will to insist anymore. You may leave it as is, if you wish.

LeonidBeynenson replied:
Thank you

druzhkov-paul commented on tools/export.py:
```diff
@@ -204,6 +207,14 @@ def main(args):
     # BEGIN nncf part
+    alt_ssd_export = getattr(args, 'alt_ssd_export', False)
+    if (is_checkpoint_nncf(args.checkpoint) and not cfg.get('nncf_config')
+        and not alt_ssd_export):
```
Why does `alt_ssd_export` affect config reading part?

LeonidBeynenson replied:
Because sometimes in our tests export.py is called for nncf-compressed model.
Just because in our OTE tests alt_ssd_export is called for all models, even for NNCF-based (and it is hard to distinct is the model NNCF or not in that point)

If alt_ssd_export is called with nncf-model it is ok, if we do not wrap it by NNCF (although most probably quality is bad, but nobody cares)

druzhkov-paul replied:
So, what will happen if we omit this condition here?

LeonidBeynenson replied:
The model will be wrapped by NNCFNetwork class, then openvino export will started with alt_ssd_export=True, and then the following assert will happen:
```python
        assert isinstance(model, detectors.SingleStageDetector)
```
-- it will through exception, because we will have NNCFNetwork model which wrap the original model, but it is not an instance of `SingleStageDetector`.

(And obviously if we remove this assert something wrong will be later, since we did not spend time to make work NNCFNetwork-wrapping with this branch of the code.)

druzhkov-paul replied:
Cool. Does not that mean, that `alt_ssd_export` is incompatible with export of NNCF-compressed model? If so you need to check it explicitly and handle the case when nncf config is already a part of config.

LeonidBeynenson replied:
Oh, I see.
I will have to try to fix that on 2 levels above, in tests of OTE, not to start alt_ssd_export for NNCF-compressed models

druzhkov-paul replied:
It's good to have this check here as well.

LeonidBeynenson replied:
Fixed!

druzhkov-paul commented on mmdet/integration/nncf/compression.py:
```diff
@@ -61,6 +64,30 @@ def is_checkpoint_nncf(path):
 
+def get_nncf_config_from_meta(path):
+    logger = get_root_logger()
+    checkpoint = torch.load(path)
```
Is there any way that does not require checkpoint reloading all the time?

LeonidBeynenson replied:
I did not think that this is a big issue.

Anyway, final loading of the model should be done _after_ the model is converted by NNCFNetwork, and if we should make this conversion (and how to make this conversion!) is detected by the configuration parameters. Now the parameters may be loaded from the meta of the checkpoint.

druzhkov-paul replied:
Loading a huge file several times is not the thing that you want to do, usually.

LeonidBeynenson replied:
Agree. Could I write TODO here? I promise to fix this -- I will add this to my special TODO list.

druzhkov-paul replied:
Yes.

druzhkov-paul commented on mmdet/integration/nncf/compression.py:
```diff
@@ -61,6 +64,30 @@ def is_checkpoint_nncf(path):
+            'get_nncf_config_from_meta should be run for NNCF-compressed checkpoints only'
+
+    config_text = meta['config']
+    with tempfile.NamedTemporaryFile(prefix='config_', suffix='.py', mode='w') as f_tmp:
+        f_tmp.write(config_text)
+        cfg = mmcv.Config.fromfile(f_tmp.name)
```
I guess it's used for debug purposes only. If so, I'd suggest adding a flag to make it only when needed.

LeonidBeynenson replied:
No. It is really required every time when we try to export an NNCF-compressed model
Otherwise export won't work

druzhkov-paul replied:
Does that mean, that you can't compose config directly from the `config_text` and have nothing to do but dump it to a file and load back?

LeonidBeynenson replied:
Yes, AFAIK it is impossible in the current mmcv -- see the implementation of `mmcv.Config`-s constructor in the file
mmcv/utils/config.py:
https://github.com/open-mmlab/mmcv/blob/master/mmcv/utils/config.py#L268
(cfg_text param of `__init__` is used only for text representation)

druzhkov-paul replied:
Ok, then...

LeonidBeynenson committed:
Fix error with CPU-only export

The error happened when a model was checked on NNCF meta info
if the model was trained on CUDA.

LeonidBeynenson committed:
Make tools/export.py raise exception if NNCF-compressed model is exported with --alt_ssd_export

LeonidBeynenson commented:
2 @druzhkov-paul : Pavel, I have fixed your comments, could you approve this PR if everything is OK?

LeonidBeynenson commented:
run tests
